### PR TITLE
feature: splat init checks for pre registrations

### LINKED ISSUE #331: Using DryIoC with an ILogger
Hello! 

We've been using Splat's logging and service location features for some time now and are tying to transition to DryIoC using the Splat.DryIoC nuget that replaces Splat's dependency resolver with a DryIoC container. 

We're having a problem where Splat's logging feature can't find the `ILogger` that we registered with the DryIoC container. For example:

```
	var c = new Container();
	c.Register<ILogger, MyLogger>();
	c.UseDryIocDependencyResolver();
	var d = Splat.Locator.Current.GetService<ILogger>();
	Splat.LogHost.Default.Debug("HELLO");
```

In the above code I would have expected that `d` would be assigned an instance of `MyLogger`, but instead it is of type `Splat.DebugLogger`. In addition (and consequently), the call to `Debug()` is routed to the `DebugLogger` instead of to `MyLogger`.

Is it possible to use both Splat.DryIoC and Splat's logging feature?

Thanks!
-Tom B.

### PR DESCRIPTION
**Pull request by dpvreony**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Resolves #331
I've changed the splat init logic to check if the resolver it is being hooked up to has already got registrations to types we're interested in (LogManager, Logger, BitmapLoader)

**What is the current behavior? (You can also link to an open issue here)**
depending on the nuance of the container involved. logging would be overridden or appended with no guaranteed order of retrieval.

**What is the new behavior (if this is a feature change)?**
tried to apply some consistency between dryioc and moderndependencyresolver. developers still have control within dryioc etc to get undesired behaviour. added documentation to that effect.

**What might this PR break?**
I've added a HasRegistration method call to the Mutable Resolver interface
- [x] I've changed DryIoc to replace registrations instead of append - can someone sense check this please :)

**Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 1]
```diff
@@ -49,6 +49,12 @@ protected ModernDependencyResolver(Dictionary<Tuple<Type, string>, List<Func<obj
             _callbackRegistry = new Dictionary<Tuple<Type, string>, List<Action<IDisposable>>>();
         }
 
+        /// <inheritdoc />
+        public bool HasRegistration(Type serviceType)
+        {
+            return _registry.Any(x => x.Key.Item1 == serviceType);
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]: You might want to change this to use the Dictionary to check. I don't think Any() is dictionary aware. 

Eg a simple `return _registry.TryGetValue(serviceType, _);` should get similar results with O(1) results.

File: src/Splat/ServiceLocation/DependencyResolverMixins.cs [hunk 2]
```diff
@@ -162,13 +162,38 @@ public static void UnregisterAll<T>(this IMutableDependencyResolver resolver, st
...
+
+        private static void RegisterDefaultLogManager(IMutableDependencyResolver resolver)
+        {
+            if (!resolver.HasRegistration(typeof(ILogManager)))
+            {
+                resolver.Register(() => new DefaultLogManager(), typeof(ILogManager));
+            }
+        }
+
+        private static void RegisterLogger(IMutableDependencyResolver resolver)
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/DependencyResolverMixins.cs]: We are eventually planning on changing the init so splitting this up makes that work easier.

eg going for a similar approach as the asp.net host extensions/builder.

File: src/Splat.SimpleInjector/SimpleInjectorDependencyResolver.cs [hunk 3]
```diff
@@ -27,12 +28,21 @@ public SimpleInjectorDependencyResolver(Container container)
...
+        }
 
         /// <inheritdoc />
         public IEnumerable<object> GetServices(Type serviceType, string contract = null) =>
             _container.GetAllInstances(serviceType);
 
+        /// <inheritdoc />
+        public bool HasRegistration(Type serviceType)
+        {
+            return _container.GetCurrentRegistrations().Any(x => x.ServiceType == serviceType);
```
Comment (glennawatson) [inline src/Splat.SimpleInjector/SimpleInjectorDependencyResolver.cs]: Could you use `_container.GetRegistration(serviceType)`?

Comment (dpvreony) [reply src/Splat.SimpleInjector/SimpleInjectorDependencyResolver.cs]: get registration made simpleinjector immutable like the issue in #240

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs (hunk 1 above)
Comment (dpvreony) [reply src/Splat/ServiceLocation/ModernDependencyResolver.cs]: cheers. will tweak later. about to pop out

File: src/Splat.SimpleInjector/SimpleInjectorDependencyResolver.cs (hunk 3 above)
Comment (glennawatson) [reply src/Splat.SimpleInjector/SimpleInjectorDependencyResolver.cs]: ah... oh well makes sense.

File: src/Splat/ServiceLocation/DependencyResolverMixins.cs (hunk 2 above)
Comment (dpvreony) [reply src/Splat/ServiceLocation/DependencyResolverMixins.cs]: ok. i'll do some more housekeeping on this

File: src/Splat.DryIoc/DryIocDependencyResolver.cs [hunk 4]
```diff
@@ -39,16 +40,22 @@ public DryIocDependencyResolver(IContainer container = null)
...
+            return _container.GetServiceRegistrations().Any(x => x.ServiceType == serviceType);
+        }
+
         /// <inheritdoc />
         public virtual void Register(Func<object> factory, Type serviceType, string contract = null)
         {
             if (string.IsNullOrEmpty(contract))
             {
-                _container.UseInstance(serviceType, factory(), IfAlreadyRegistered.AppendNewImplementation);
+                _container.UseInstance(serviceType, factory(), IfAlreadyRegistered.Replace);
```
Comment (glennawatson) [inline src/Splat.DryIoc/DryIocDependencyResolver.cs]: This won't match the modern dependency resolver I don't think, it just appends to the linked list for that bucket. Not sure if that matters.

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs (hunk 1 above)
Comment (dpvreony) [reply src/Splat/ServiceLocation/ModernDependencyResolver.cs]: adjusting to

`
        public bool HasRegistration(Type serviceType)
        {
            var pair = Tuple.Create(serviceType, string.Empty);
            return _registry.ContainsKey(pair);
        }
`

File: src/Splat.DryIoc/DryIocDependencyResolver.cs (hunk 4 above)
Comment (dpvreony) [reply src/Splat.DryIoc/DryIocDependencyResolver.cs]: doh when i was comparing the behaviours, i missed the ! in the modern dependency resolver

`if (!_registry.ContainsKey(pair))`

modern dependency resolver returns last item in the bucket.
1) do we need to hold multiple registrations of things like the ilogger factory? if it's being done by type and contractname?
2) other ioc containers may not return the last item (if they even support multiple registrations) so we have to be aware of nuances. The logging extensions may not work as desired on anything other than modern dependency resolver.

i'll have a think and build some tests up. welcome thoughts as i only use very lightweight use cases of the servicelocator.

**Comment by glennawatson** [comment]
I been slowly merging everything over to valuetuple format over tuple. Usually has greater self documentation also performance advantages.

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs (hunk 1 above)
Comment (dpvreony) [reply src/Splat/ServiceLocation/ModernDependencyResolver.cs]: now using ValueTuple

File: src/Splat.DryIoc/DryIocDependencyResolver.cs (hunk 4 above)
Comment (dpvreony) [reply src/Splat.DryIoc/DryIocDependencyResolver.cs]: autofac, dryioc, modern dependency resolver now have same behaviour.

File: src/Splat/ServiceLocation/DependencyResolverMixins.cs [hunk 5]
```diff
@@ -162,13 +162,38 @@ public static void UnregisterAll<T>(this IMutableDependencyResolver resolver, st
...
+        {
+            if (!resolver.HasRegistration(typeof(ILogger)))
+            {
+                resolver.RegisterConstant(new DebugLogger(), typeof(ILogger));
```
Comment (dpvreony) [inline src/Splat/ServiceLocation/DependencyResolverMixins.cs]: need to think about this. it never changes when we change a logmanager

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 6]
```diff
@@ -40,19 +40,26 @@ public ModernDependencyResolver()
         /// Initializes a new instance of the <see cref="ModernDependencyResolver"/> class.
         /// </summary>
         /// <param name="registry">A registry of services.</param>
-        protected ModernDependencyResolver(Dictionary<Tuple<Type, string>, List<Func<object>>> registry)
+        protected ModernDependencyResolver(Dictionary<ValueTuple<Type, string>, List<Func<object>>> registry)
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]: Dictionary<(Type type, string typeName), List<Func<object>>>>

Comment (glennawatson) [reply src/Splat/ServiceLocation/ModernDependencyResolver.cs]: That's the C# shorthand syntax for ValueTuple.

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 7]
```diff
@@ -40,19 +40,26 @@ public ModernDependencyResolver()
...
+        {
+            var pair = ValueTuple.Create(serviceType, string.Empty);
+            return _registry.ContainsKey(pair);
         }
 
         /// <inheritdoc />
         public void Register(Func<object> factory, Type serviceType, string contract = null)
         {
-            var pair = Tuple.Create(serviceType, contract ?? string.Empty);
+            var pair = ValueTuple.Create(serviceType, contract ?? string.Empty);
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]:
```suggestion
            var pair = (serviceType, contract ?? string.Empty);
```

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 8]
```diff
@@ -94,7 +101,7 @@ public void Register(Func<object> factory, Type serviceType, string contract = n
         /// <inheritdoc />
         public object GetService(Type serviceType, string contract = null)
         {
-            var pair = Tuple.Create(serviceType, contract ?? string.Empty);
+            var pair = ValueTuple.Create(serviceType, contract ?? string.Empty);
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]:
```suggestion
            var pair = (serviceType, contract ?? string.Empty);
```

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 9]
```diff
@@ -107,7 +114,7 @@ public object GetService(Type serviceType, string contract = null)
         /// <inheritdoc />
         public IEnumerable<object> GetServices(Type serviceType, string contract = null)
         {
-            var pair = Tuple.Create(serviceType, contract ?? string.Empty);
+            var pair = ValueTuple.Create(serviceType, contract ?? string.Empty);
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]:
```suggestion
            var pair = (serviceType, contract ?? string.Empty);
```

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 10]
```diff
@@ -119,28 +126,34 @@ public IEnumerable<object> GetServices(Type serviceType, string contract = null)
         /// <inheritdoc />
         public void UnregisterCurrent(Type serviceType, string contract = null)
         {
-            var pair = Tuple.Create(serviceType, contract ?? string.Empty);
+            var pair = ValueTuple.Create(serviceType, contract ?? string.Empty);
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]:
```suggestion
            var pair = (serviceType, contract ?? string.Empty);
```

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 11]
```diff
@@ -119,28 +126,34 @@ public IEnumerable<object> GetServices(Type serviceType, string contract = null)
...
+            }
+
+            list.RemoveAt(position);
         }
 
         /// <inheritdoc />
         public void UnregisterAll(Type serviceType, string contract = null)
         {
-            var pair = Tuple.Create(serviceType, contract ?? string.Empty);
+            var pair = ValueTuple.Create(serviceType, contract ?? string.Empty);
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]:
```suggestion
            var pair = (serviceType, contract ?? string.Empty);
```

File: src/Splat/ServiceLocation/ModernDependencyResolver.cs [hunk 12]
```diff
@@ -119,28 +126,34 @@ public IEnumerable<object> GetServices(Type serviceType, string contract = null)
...
 
             _registry[pair] = new List<Func<object>>();
         }
 
         /// <inheritdoc />
         public IDisposable ServiceRegistrationCallback(Type serviceType, string contract, Action<IDisposable> callback)
         {
-            var pair = Tuple.Create(serviceType, contract ?? string.Empty);
+            var pair = ValueTuple.Create(serviceType, contract ?? string.Empty);
```
Comment (glennawatson) [inline src/Splat/ServiceLocation/ModernDependencyResolver.cs]:
```suggestion
            var pair = (serviceType, contract ?? string.Empty);
```

**Comment by glennawatson** [comment]
This is a good overview and how to use the new value tuple shorthand https://blogs.msdn.microsoft.com/mazhou/2017/05/26/c-7-series-part-1-value-tuples/

File: src/Splat.DryIoc/DryIocDependencyResolver.cs (hunk 4 above)
Comment (glennawatson) [reply src/Splat.DryIoc/DryIocDependencyResolver.cs]: There is a method associated with GetAllRegistrations too that might need some testing around for the various DI engines to make sure they work as we expect.

Comment (glennawatson) [reply src/Splat.DryIoc/DryIocDependencyResolver.cs]: https://github.com/reactiveui/splat/blob/6f3381952f3680abb8d718eda974a3d4fd53d052/src/Splat/ServiceLocation/IReadonlyDependencyResolver.cs#L32

**Comment by glennawatson** [comment]
I just did a benchmark of using the ValueTuple directly vs GetKey()

```
|         Method |     Mean |     Error |    StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|---------:|------:|------:|------:|----------:|
|    TestWithGet | 1.198 ms | 0.0423 ms | 0.1201 ms | 1.164 ms |     - |     - |     - |         - |
| TestWithoutGet | 1.179 ms | 0.0344 ms | 0.0954 ms | 1.156 ms |     - |     - |     - |         - |
```

My test was simple
```
        [Benchmark]
        public void TestWithGet()
        {
            for (int i = 0; i < NumberCases; ++i)
            {
                var key = Get(ElementsNames1[i], ElementsNames2[i]);

                var element = _indexToElements[key];

                outputElements.Add(element);
            }
        }

        [Benchmark]
        public void TestWithoutGet()
        {
            for (int i = 0; i < NumberCases; ++i)
            {
                var key = (ElementsNames1[i], ElementsNames2[i]);

                var element = _indexToElements[key];

                outputElements.Add(element);
            }
        }
```

That's with randomly allocated keys etc.

So that confirms the way you are doing it is fractionally slower if anything meaningful.

**Comment by dpvreony** [comment]
should read c#7... anyway I've tweaked

> Also just did this for curiosity, added the following benchmark
>         [Benchmark]
>         public void TestWithTupleGet()
>         {
>             for (int i = 0; i < NumberCases; ++i)
>             {
>                 var key = Tuple.Create(ElementsNames1[i], ElementsNames2[i]);
> 
>                 var element = _indexToTupleElements[key];
> 
>                 outputElements.Add(element);
>             }
>         }
> |           Method |     Mean |     Error |    StdDev |   Median | Gen 0 | Gen 1 | Gen 2 | Allocated |
> |----------------- |---------:|----------:|----------:|---------:|------:|------:|------:|----------:|
> |      TestWithGet | 1.297 ms | 0.0626 ms | 0.1754 ms | 1.240 ms |     - |     - |     - |         - |
> |   TestWithoutGet | 1.361 ms | 0.0837 ms | 0.2401 ms | 1.278 ms |     - |     - |     - |         - |
> | TestWithTupleGet | 2.212 ms | 0.0883 ms | 0.2371 ms | 2.194 ms |     - |     - |     - |  320000 B |

With and Without both inside the margin for error? And the result reversed 2nd time around. I can decorate with MethodImplOptions.AggressiveInlining? (if it's not already inlining?) or put it back. I've no preference :)

**Comment by glennawatson** [comment]
The error increased on the second run. Which means my computer was probably under greater strain at the time.

**Comment by glennawatson** [comment]
I think that shows how little margin is in it that the Get() approach is fine, but getting rid of Tuple is validated.

**Comment by glennawatson** [comment]
Might as well bump the minor version to one by incrementing the second number in version.json in the root directory for this one.

**Comment by dpvreony** [comment]
be nice if the todo comments were done as closable review, deleted them for clarity on this PR

File: src/Splat/ServiceLocation/DependencyResolverMixins.cs (hunk 5 above)
Comment (dpvreony) [reply src/Splat/ServiceLocation/DependencyResolverMixins.cs]: moved this to #367

**Comment by glennawatson** [comment]
got much left to do on this one @dpvreony ?

**Pull request merged**